src/Rules/Operators/InvalidBinaryOperationRule.php

```diff
@@ -44,81 +45,83 @@ public function processNode(Node $node, Scope $scope): array
-			} else {
-				$identifier = 'binaryOp';
-				$newNode = clone $node;
-				$newNode->setAttribute('phpstan_cache_printer', null);
-				$left = $node->left;
-				$right = $node->right;
-				$newNode->left = $leftVariable;
-				$newNode->right = $rightVariable;
-			}
-
-			if ($node instanceof Node\Expr\AssignOp\Concat || $node instanceof Node\Expr\BinaryOp\Concat) {
-				$callback = static fn (Type $type): bool => !$type->toString() instanceof ErrorType;
-			} else {
-				$callback = static fn (Type $type): bool => !$type->toNumber() instanceof ErrorType;
-			}
-
-			$leftType = $this->ruleLevelHelper->findTypeToCheck(
-				$scope,
-				$left,
-				'',
-				$callback,
-			)->getType();
-			if ($leftType instanceof ErrorType) {
-				return [];
-			}
-
-			$rightType = $this->ruleLevelHelper->findTypeToCheck(
-				$scope,
-				$right,
-				'',
-				$callback,
-			)->getType();
-			if ($rightType instanceof ErrorType) {
-				return [];
-			}
-
-			if (!$scope instanceof MutatingScope) {
-				throw new ShouldNotHappenException();
-			}
-
-			$scope = $scope
-				->assignVariable($leftName, $leftType, $leftType)
-				->assignVariable($rightName, $rightType, $rightType);
-
-			if (!$scope->getType($newNode) instanceof ErrorType) {
-				return [];
-			}
-
-			return [
-				RuleErrorBuilder::message(sprintf(
-					'Binary operation "%s" between %s and %s results in an error.',
-					substr(substr($this->exprPrinter->printExpr($newNode), strlen($leftName) + 2), 0, -(strlen($rightName) + 2)),
-					$scope->getType($left)->describe(VerbosityLevel::value()),
-					$scope->getType($right)->describe(VerbosityLevel::value()),
-				))
-					->line($left->getStartLine())
-					->identifier(sprintf('%s.invalid', $identifier))
-					->build(),
-			];
+		if (!$scope->getType($node) instanceof ErrorType && !$this->bleedingEdge) {
+			return [];
+		}
+
+		$leftName = '__PHPSTAN__LEFT__';
+		$rightName = '__PHPSTAN__RIGHT__';
+		$leftVariable = new Node\Expr\Variable($leftName);
+		$rightVariable = new Node\Expr\Variable($rightName);
+		if ($node instanceof Node\Expr\AssignOp) {
+			$identifier = 'assignOp';
+			$newNode = clone $node;
+			$newNode->setAttribute('phpstan_cache_printer', null);
+			$left = $node->var;
+			$right = $node->expr;
+			$newNode->var = $leftVariable;
+			$newNode->expr = $rightVariable;
+		} else {
+			$identifier = 'binaryOp';
+			$newNode = clone $node;
+			$newNode->setAttribute('phpstan_cache_printer', null);
+			$left = $node->left;
+			$right = $node->right;
+			$newNode->left = $leftVariable;
+			$newNode->right = $rightVariable;
+		}
+
+		if ($node instanceof Node\Expr\AssignOp\Concat || $node instanceof Node\Expr\BinaryOp\Concat) {
+			$callback = static fn (Type $type): bool => !$type->toString() instanceof ErrorType;
+		} elseif ($node instanceof Node\Expr\AssignOp\Plus || $node instanceof Node\Expr\BinaryOp\Plus) {
+			$callback = static fn (Type $type): bool => !$type->toNumber() instanceof ErrorType || $type->isArray()->yes();
```

Comment:
This is new, otherwise the code in this file should be the same with the exception of the early exit and not requiring `$scope->getType($node) instanceof ErrorType` anymore in bleeding edge.